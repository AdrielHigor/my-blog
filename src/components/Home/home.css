#post-root{
    margin:auto;
    width: auto;
    max-width: 1080px;
}

a {
    color: #000000;
    text-decoration: none;
}

article#post{
    margin: 3px auto;
    padding: 20px;
    background-color: #e8e8e8;
    min-height: 150px;
    border-style: solid;
    border-width: 1px;
    border-radius: 3px;
    border-color: #d1d1d1;
}

article#post:hover{
    background-color: #d6d5d5;
}

div#title{
    display: flex;
    flex-direction: column;
}

img#post-image {
    width: 100%;
    height: auto;
    padding: 10px;
}

strong#post-title {
    padding: 10px;
    font-size: 40px;
    text-align: left;
}

span#post-author {
    padding: 10px;
    font-size: 16px;
}

p#post-description {
    padding-left: 10px;
    font-size: 18px;
    overflow: hidden;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
}
