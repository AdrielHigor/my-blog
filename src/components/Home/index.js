import React, { Component, useState, useEffect } from "react";
import "./home.css";
import firebase from "../../firebase";
import { Link } from "react-router-dom";

export default class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      posts: []
    };
  }

  componentDidMount() {
    let postsArray = [];

    firebase.app.ref("posts").on("child_added", snapshot => {
      postsArray.push({
        key: snapshot.key,
        description: snapshot.val().description,
        author: snapshot.val().author,
        title: snapshot.val().title,
        image: snapshot.val().image
      });
      postsArray.reverse();
      this.setState({ posts: postsArray });
    });
  }

  componentDidUpdate() {
    firebase.app.ref("posts").on("child_removed", snapshot => {
      this.setState(prevState => ({
        posts: prevState.posts.filter(item => item.key !== snapshot.key)
      }));
    });
  }

  componentWillUnmount() {
    firebase.app.ref("posts").off("child_changed");
    firebase.app.ref("posts").off("child_removed");
  }

  render() {
    return (
      <section id="post-root">
        {this.state.posts.map(post => {
          return (
            <Link to={"/post/" + post.key} key={post.key}>
              <article id="post">
                <header>
                  <div id="title">
                    <strong id="post-title">{post.title}</strong>
                    <span id="post-author">Autor: {post.author}</span>
                  </div>
                </header>
                <img id="post-image" src={post.image} alt="" />
                <footer>
                  <p id="post-description">{post.description}</p>
                </footer>
              </article>
            </Link>
          );
        })}
      </section>
    );
  }
}
