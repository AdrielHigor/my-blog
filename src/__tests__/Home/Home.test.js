import React from "react";
import Home from "../../components/Home";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import "../../setupTests";

describe("Home Test Suite", () => {
    let component

    jest.mock('../../firebase', () => {
        return {
            app: {
                ref: jest.fn(() => ({
                    on: jest.fn()
                }))
            }
        };
    });

    beforeEach(() => {
        component = shallow(<Home />)
        component.setState({ posts: [] })
    })

    test("Home rendering correctly", () => {
        expect(component.html()).toMatchSnapshot();
    })

    test("Home with empty posts state", () => {
        expect(component.find("section").children().length).toEqual(0);
        expect(component.find(Link).length).toEqual(0);
    })

    test("Home with data in posts state", () => {
        component.setState({
            posts: [{
                key: 1,
                description: "post",
                author: "Test Author",
                title: "Test Post",
                image: "http://testimage.com/test_image.jpg",
            }]
        })

        expect(component.find("section").children().length).toBeGreaterThan(0);
        expect(component.find(Link).length).toBeGreaterThan(0);
    })
})
