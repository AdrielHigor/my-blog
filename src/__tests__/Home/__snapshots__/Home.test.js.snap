// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Test Suite Home rendering correctly 1`] = `"<section id=\\"post-root\\"></section>"`;
